### PR TITLE
Add template methods around DbConnection inside RelationalConnection

### LINKED ISSUE #21327: Consider adding template methods for DbConnection calls
For #7283 I need to call a different overload of DbConnection.Open in the SQL Server provider. I'm adding a template method to allow this, but similar template methods could be useful in other providers. Tentative list of relevant methods based on what RelationalCommand calls:

* BeginTransaction
* Close
* EnlistTransaction
* State
* Dispose

We could also consider doing this for other Db types where they are wrapped by EF.

### PR DESCRIPTION
**Pull request by joaopgrassi**
Resolves #21327

### CODE REVIEW

**Comment by joaopgrassi** [comment]
There's a test failing (`RelationalApiConsistencyTest`) due to the newly template method for `DbConnection.CloseAsync()`... Not sure how to proceed. The method is protected, so I can't add to the exclusion list (`AsyncMethodExceptions`). If I make it public then it's not consistent with the rest of the template methods.. and just adding the cancellation token to make the tests pass seems hacky (since it's not used by `DbConnection.CloseAsync`)

File: src/EFCore.Relational/Storage/RelationalConnection.cs [hunk 1]
```diff
@@ -227,11 +227,19 @@ public virtual void EnlistTransaction(Transaction transaction)
                 Dependencies.TransactionLogger.ExplicitTransactionEnlisted(this, transaction);
             }
 
-            DbConnection.EnlistTransaction(transaction);
+            ConnectionEnlistTransaction(transaction);
```
Comment (ajcvickers) [inline src/EFCore.Relational/Storage/RelationalConnection.cs:230]: @dotnet/efteam Any thoughts on naming these template methods before merging? (We will review in API review later anyway.)

**Comment by ajcvickers** [comment]
@joaopgrassi You should be able to add the method to `AsyncMethodExceptions` even though it is protected. Something like:

```C#
typeof(RelationalConnection).GetMethod("CreateDbConnection", BindingFlags.Instance | BindingFlags.NonPublic);
```

**Comment by joaopgrassi** [comment]
>typeof(RelationalConnection).GetMethod("CreateDbConnection", BindingFlags.Instance | BindingFlags.NonPublic);

Right 😅. Was so deep into the errors forgot the obvious, lol. 

I did notice though that some methods return a `ValueTask` but they were not being caught by this consistency check. Looking more I saw that's because inside `Async_methods_should_have_overload_with_cancellation_token_and_end_with_async_suffix` it just looks for `Task`. Not sure if it's something to change as well.

**Review by ajcvickers — APPROVED** [review]

**Comment by ajcvickers** [comment]
@joaopgrassi Thanks for the contribution!

**Pull request merged**